homeassistant/components/homematicip_cloud/light.py

```diff
@@ -258,6 +275,108 @@ async def async_turn_off(self, **kwargs: Any) -> None:
         )
 
 
+class HomematicipNotificationLightV2(HomematicipGenericEntity, LightEntity):
```

Comment:
A lot of code is duplicated from `HomematicipNotificationLight`. Can we not extend the `HomematicipNotificationLight` instead?

Reply:
Both have a different properties like effects, features, calling different functions. I would be prefer to separate them in different classes.

Reply:
There are still a lot of common code.
The code can be simplified by
```diff
diff --git a/homeassistant/components/homematicip_cloud/light.py b/homeassistant/components/homematicip_cloud/light.py
index 2b39f4c3b6..978a9c4878 100644
--- a/homeassistant/components/homematicip_cloud/light.py
+++ b/homeassistant/components/homematicip_cloud/light.py
@@ -50,28 +50,15 @@ async def async_setup_entry(
         elif isinstance(device, AsyncBrandSwitchNotificationLight):
             device_version = Version(device.firmwareVersion)
             entities.append(HomematicipLight(hap, device))
+            entity_class = HomematicipNotificationLight
             if device_version > Version("2.0.0"):
-                entities.append(
-                    HomematicipNotificationLightV2(
-                        hap, device, device.topLightChannelIndex, "Top"
-                    )
-                )
-                entities.append(
-                    HomematicipNotificationLightV2(
-                        hap, device, device.bottomLightChannelIndex, "Bottom"
-                    )
-                )
-            else:
-                entities.append(
-                    HomematicipNotificationLight(
-                        hap, device, device.topLightChannelIndex
-                    )
-                )
-                entities.append(
-                    HomematicipNotificationLight(
-                        hap, device, device.bottomLightChannelIndex
-                    )
-                )
+                entity_class = HomematicipNotificationLightV2
+            entities.append(
+                entity_class(hap, device, device.topLightChannelIndex, "Top")
+            )
+            entities.append(
+                entity_class(hap, device, device.bottomLightChannelIndex, "Bottom")
+            )
         elif isinstance(device, (AsyncWiredDimmer3, AsyncDinRailDimmer3)):
             entities.extend(
                 HomematicipMultiDimmer(hap, device, channel=channel)
@@ -174,16 +161,9 @@ class HomematicipNotificationLight(HomematicipGenericEntity, LightEntity):
     _attr_supported_color_modes = {ColorMode.HS}
     _attr_supported_features = LightEntityFeature.TRANSITION
 
-    def __init__(self, hap: HomematicipHAP, device, channel: int) -> None:
+    def __init__(self, hap: HomematicipHAP, device, channel: int, post: str) -> None:
         """Initialize the notification light entity."""
-        if channel == 2:
-            super().__init__(
-                hap, device, post="Top", channel=channel, is_multi_channel=True
-            )
-        else:
-            super().__init__(
-                hap, device, post="Bottom", channel=channel, is_multi_channel=True
-            )
+        super().__init__(hap, device, post=post, channel=channel, is_multi_channel=True)
 
         self._color_switcher: dict[str, tuple[float, float]] = {
             RGBColorState.WHITE: (0.0, 0.0),
@@ -275,11 +255,10 @@ class HomematicipNotificationLight(HomematicipGenericEntity, LightEntity):
         )
 
 
-class HomematicipNotificationLightV2(HomematicipGenericEntity, LightEntity):
+class HomematicipNotificationLightV2(HomematicipNotificationLight):
     """Representation of HomematicIP Cloud notification light."""
 
-    _attr_color_mode = ColorMode.HS
-    _attr_supported_color_modes = {ColorMode.HS}
+    _attr_supported_features = LightEntityFeature.EFFECT
     _effect_list = [
         OpticalSignalBehaviour.BILLOW_MIDDLE,
         OpticalSignalBehaviour.BLINKING_MIDDLE,
@@ -288,36 +267,6 @@ class HomematicipNotificationLightV2(HomematicipGenericEntity, LightEntity):
         OpticalSignalBehaviour.ON,
     ]
 
-    def __init__(self, hap: HomematicipHAP, device, channel: int, post: str) -> None:
-        """Initialize the notification light entity."""
-        super().__init__(hap, device, post=post, channel=channel, is_multi_channel=True)
-
-        self._attr_supported_features |= LightEntityFeature.EFFECT
-        self._color_switcher: dict[str, tuple[float, float]] = {
-            RGBColorState.WHITE: (0.0, 0.0),
-            RGBColorState.RED: (0.0, 100.0),
-            RGBColorState.YELLOW: (60.0, 100.0),
-            RGBColorState.GREEN: (120.0, 100.0),
-            RGBColorState.TURQUOISE: (180.0, 100.0),
-            RGBColorState.BLUE: (240.0, 100.0),
-            RGBColorState.PURPLE: (300.0, 100.0),
-        }
-
-    @property
-    def _func_channel(self) -> NotificationLightChannel:
-        return self._device.functionalChannels[self._channel]
-
-    @property
-    def brightness(self) -> int:
-        """Return the brightness of this light between 0..255."""
-        return int((self._func_channel.dimLevel or 0.0) * 255)
-
-    @property
-    def hs_color(self) -> tuple[float, float]:
-        """Return the hue and saturation color value [float, float]."""
-        simple_rgb_color = self._func_channel.simpleRGBColorState
-        return self._color_switcher.get(simple_rgb_color, (0.0, 0.0))
-
     @property
     def effect_list(self) -> list[str] | None:
         """Return the list of supported effects."""
@@ -333,21 +282,6 @@ class HomematicipNotificationLightV2(HomematicipGenericEntity, LightEntity):
         """Return true if light is on."""
         return self._func_channel.on
 
-    @property
-    def extra_state_attributes(self) -> dict[str, Any]:
-        """Return the state attributes of the notification light sensor."""
-        state_attr = super().extra_state_attributes
-
-        if self.is_on:
-            state_attr[ATTR_COLOR_NAME] = self._func_channel.simpleRGBColorState
-
-        return state_attr
-
-    @property
-    def unique_id(self) -> str:
-        """Return a unique ID."""
-        return f"{self.__class__.__name__}_{self._post}_{self._device.id}"
-
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Turn the light on."""
         # Use hs_color from kwargs,
```

Reply:
@hahn-th is there a reason you don't want to do this deduplication of code?

Reply:
@emontnemery just because I am quite busy at work and tired in the evening... sorry about that. I will take care of it this weekend or next week

Reply:
Take your time, I just saw the PR is marked as ready for review but the suggestion for deduplicating was not addressed.
I've set the PR to draft now, please don't forget to click the "Ready for review" button once you've done the refactoring.

Reply:
Hey @edenhaus @emontnemery. I found the time to fix the redundant code. Sorry for delay.